hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.metrics;
+
+import com.hazelcast.internal.metrics.MetricTagger;
+import com.hazelcast.internal.metrics.MetricsCollectionContext;
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.internal.metrics.ProbeUnit;
+import com.hazelcast.jet.core.metrics.Metric;
+import com.hazelcast.jet.core.metrics.MetricTags;
+import com.hazelcast.jet.core.metrics.Unit;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.BiFunction;
+
+public class MetricsContext {
+
+    private String onlyName;
+    private Metric onlyMetric;
+
+    private Map<String, Metric> metrics;
+
+    Metric metric(String name, Unit unit) {
+        return metric(name, unit, SimpleMetric::new);
+    }
+
+    Metric threadSafeMetric(String name, Unit unit) {
+        return metric(name, unit, ThreadSafeMetric::new);
+    }
+
+    private Metric metric(String name, Unit unit, BiFunction<String, Unit, Metric> metricSupplier) {
+        if (metrics == null) { //at most one already defined metric
+            if (onlyMetric == null) { //no already defined metrics
+                onlyMetric = metricSupplier.apply(name, unit);
+                onlyName = name;
+                return onlyMetric;
+            } else { //one single already defined metric
+                if (name.equals(onlyName)) { //single already defined metric same as the requested one
+                    return onlyMetric;
+                } else { //single already defined metric different from the requested one
+                    metrics = new HashMap<>();
+                    metrics.put(onlyName, onlyMetric);
+
+                    onlyMetric = null;
+                    onlyName = null;
+
+                    Metric metric = metricSupplier.apply(name, unit);
+                    metrics.put(name, metric);
+                    return metric;
+                }
+            }
+        } else { //multiple metrics already defined
+            Metric metric = metrics.get(name);
+            if (metric == null) { //requested metric not yet defined
+                metric = metricSupplier.apply(name, unit);
```

Comment:
This branch is not covered by tests

Reply:
Hmm... I fixed the coverage, but there is a problem... So I've added a new metric to `MetricsTest.counter`, called "last", like this:

                .drawFrom(TestSources.items(0L, 1L, 2L, 3L, 4L))
                .filter(l -> {
                    boolean pass = l % 2 == 0;

                    if (!pass) {
                        Metrics.metric("dropped").increment();
                    }
                    Metrics.metric("total").increment();
                    System.out.println("l = " + l); //todo: remove
                    Metrics.metric("last").set(l);

                    return pass;
                })

I was naively expecting a value of 4 or at least one of the input values... But it's far from it, the result is 10, because each input goes to a different processor and the various "last" values from the multiple processors get summed up.

This behaviour is ok, when we count things and use the increment and decrement operations, but with this "set" operation and other such intentions summing up the values per processor is not really what we need... 

What do you think we should do? Maybe implement some sort of option for metric where we specify how they should be aggregated?